File: samples/kaleido-sample-ecommerce-ui/src/app/kaleido/models/queryable-request.ts
import { SortDirection } from './sort-direction';
import { FilterOperator, LogicalOperator } from './enumerations';

//export interface QueryRequest {
//  query: QueryBody;
//}

export interface QueryRequest<TParameters = undefined> {
  parameters?: TParameters;
  query?: QueryBody;
}

export interface QueryBody {
  searchText?: string;
  filter?: QueryFilterNode;
  sort?: QuerySort[];
  page?: QueryPage;
}

export interface QueryPage {
  size: number;
  offset: number;
}

export interface QuerySort {
    field: string;
    direction: SortDirection;
    sequence: number;
}

export interface QueryFilterNode {
  condition?: QueryFilterCondition;
  group?: QueryFilterGroup;
}

export interface QueryFilterCondition {
  field: string;
  operator: FilterOperator;
  values: unknown[];
}

export interface QueryFilterGroup {
  operator: LogicalOperator;
  filters: QueryFilterNode[];
}

File: samples/PriorAuth/Intake.Artifacts/Process/Models/QueryableRequest.cs
namespace Kaleido.Samples.PriorAuth.Intake.Artifacts.Process.Models;

public sealed record QueryRequest<TParameters>
{
    public TParameters? Parameters { get; init; }

    public QueryBody? Query { get; init; }
}

public sealed record QueryBody
{
    public string? SearchText { get; init; }

    public QueryFilterNode? Filter { get; init; }
}

public sealed record QueryFilterNode
{
    public QueryFilterCondition? Condition { get; init; }

    public QueryFilterGroup? Group { get; init; }
}

public sealed record QueryFilterCondition
{
    public string Field { get; init; } = string.Empty;

    public string Operator { get; init; } = "Equals";

    public IReadOnlyCollection<object> Values { get; init; } = [];
}

public sealed record QueryFilterGroup
{
    public string Operator { get; init; } = "And";

    public IReadOnlyCollection<QueryFilterNode> Filters { get; init; } = [];
}

public static class QueryRequestFactory
{
    public static QueryRequest<object> CreateEqualsRequest(
        params (string Field, object Value)[] filters)
    {
        return new QueryRequest<object>
        {
            Query = new QueryBody
            {
                Filter = new QueryFilterNode
                {
                    Group = new QueryFilterGroup
                    {
                        Filters = filters
                            .Select(filter => new QueryFilterNode
                            {
                                Condition = new QueryFilterCondition
                                {
                                    Field = filter.Field,
                                    Values = [filter.Value]
                                }
                            })
                            .ToArray()
                    }
                }
            }
        };
    }
}
